core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java

```diff
@@ -0,0 +1,448 @@
+    private static final int SCALE_WRITERS_PARTITION_COUNT = 4096;
+    // If the percentage difference between the two different task buckets with maximum and minimum processed bytes
+    // since last rebalance is above 0.7 (or 70%), then we consider them skewed.
+    private static final double TASK_BUCKET_SKEWNESS_THRESHOLD = 0.7;
+    private static final long MIN_DATA_PROCESSED_REBALANCE_THRESHOLD = DataSize.of(50, MEGABYTE).toBytes();
+
+    private final int partitionCount;
+    private final int taskCount;
+    private final int taskBucketCount;
+    private final long minPartitionDataProcessedRebalanceThreshold;
+    private final long minDataProcessedRebalanceThreshold;
+
+    private final AtomicLongArray partitionRowCount;
+    private final AtomicLong dataProcessed;
+    private final AtomicLong dataProcessedAtLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionDataSizeAtLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionDataSizeSinceLastRebalancePerTask;
+
+    @GuardedBy("this")
+    private final long[] estimatedTaskBucketDataSizeSinceLastRebalance;
+
+    private final List<List<TaskBucket>> partitionAssignments;
+
+    public static boolean checkCanScalePartitionsRemotely(Session session, int taskCount, PartitioningScheme partitioningScheme, NodePartitioningManager nodePartitioningManager)
+    {
+        // In case of connector partitioning, check if bucketToPartitions has fixed mapping or not. If it is fixed
+        // then we can't distribute a bucket across multiple tasks.
+        Boolean hasFixedNodeMapping = nodePartitioningManager.getConnectorBucketNodeMap(session, partitioningScheme.getPartitioning().getHandle())
+                .map(ConnectorBucketNodeMap::hasFixedMapping)
+                .orElse(false);
+        // Use skewed partition rebalancer only when there are more than one tasks
+        return taskCount > 1 && !hasFixedNodeMapping && isScaledWriterHashDistribution(partitioningScheme.getPartitioning().getHandle());
+    }
+
+    public static PartitionFunction createPartitionFunction(
+            Session session,
+            NodePartitioningManager nodePartitioningManager,
+            PartitioningScheme scheme,
+            List<Type> partitionChannelTypes)
+    {
+        PartitioningHandle handle = scheme.getPartitioning().getHandle();
+        // In case of SystemPartitioningHandle we can use arbitrary bucket count so that skewness mitigation
+        // is more granular.
+        // Whereas, in the case of connector partitioning we have to use connector provided bucketCount
+        // otherwise buckets will get mapped to tasks incorrectly which could affect skewness handling.
+        //
+        // For example: if there are 2 hive buckets, 2 tasks, and 10 artificial bucketCount then this
+        // could be how actual hive buckets are mapped to artificial buckets and tasks.
+        //
+        // hive bucket  artificial bucket   tasks
```

Comment:
nit: that is one, but also if node count is 11 and bucket count is 8, then it will effectively mean that each node will write every bucket (so it's effectively round robin from the start).